File: 236/lowest-common-ancestor-of-a-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool LCA(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q, struct TreeNode** ans) {
    if(root==NULL) return false;
    bool center = (root==p || root==q);
    bool left = LCA(root->left, p, q, ans);
    bool right = LCA(root->right, p, q, ans);
    if( (left && right) || (left && center) || (right && center) ) {
        *ans = root;
//printf("root->val:%d\n", root->val);
    }
    return (left||right||center);    
}

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    struct TreeNode* ans;
    LCA(root, p, q, &ans);
    return ans;
}

File: 236/lowest-common-ancestor-of-a-binary-tree.py
# LeetCode 236. Lowest Common Ancestor of a Binary Tree
# 在 Binary tree 裡，找到 p 和 q 的共同祖先
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.ans = None
        def helper(root, p, q): # 看「中幾個」
            if self.ans != None: return 2  # 找到答案，直接結束
            if root==None: return 0  # 終止條件
            found = 0
            if root==p or root==q: found += 1
            found += helper(root.left, p, q) + helper(root.right, p, q)
            if found==2 and self.ans == None:
                self.ans = root
            return found
        helper(root, p, q)
        if self.ans==None: return root
        return self.ans

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
